example/weasel/language design.md

```diff
@@ -0,0 +1,197 @@
+    | "mul" | "*"
+    | "div" | "/"
+    | "rem" | "%"     (on wasm_int_type or integer)
+    | "min"
+    | "max"
+    | "and" | "&"     (on wasm_int_type or integer)
+    | "or" | "|"      (on wasm_int_type or integer)
+    | "xor" | "^"     (on wasm_int_type or integer)
+    | "shl" | "<<"    (on wasm_int_type or integer)
+    | "shr_s" | ">>"  (on wasm_int_type or integer)
+    | "shr_x" | ">>>" (on wasm_int_type or integer)
+    | "rotl"          (on wasm_int_type)
+    | "rotr"          (on wasm_int_type)
+    | "copysign"
+
+memarg ::= "(" "base" term ")" {"(" "offset" term ")"}? {"(" "align" term ")"}?
+
+term ::= "result"
+    | integer
+    | real
+    | i32_max
+    | i32_min
+    | i64_max
+    | i64_min
+    | ind
+    | pterm
+
+pterm ::= "result" {integer}?
+    | "old" term
+    | "param" ind
+    | "global" ind
+    | "binder" ind
+    | unop term
+    | binop term term
+    | "null" ref_type
+    | "ref_func" ind
+    | "table_get" ind
+    | "table_size" ind
+    | "i32_load" memarg
+    | "i64_load" memarg
+    | "f32_load" memarg
+    | "f64_load" memarg
+    | "i32_load8_s" memarg
+    | "i32_load8_x" memarg
+    | "i32_load16_s" memarg
+    | "i32_load16_x" memarg
+    | "i64_load8_s" memarg
+    | "i64_load8_x" memarg
+    | "i64_load16_s" memarg
+    | "i64_load16_x" memarg
+    | "i64_load32_s" memarg
+    | "i64_load32_x" memarg
+    | "memory_size" ind           (now there is only one memory)
+    | "?:" term term term
+    | "let" {id}? term term
+    | "cast" numerical_type term
+    | (* function application *)
+
+unpred ::= "eqz" | "=0?"
+    | "is_null"       (on ref_type)
+
+binpred ::= "eq" | "="
+    | "ne" | "!="
+    | "lt" | "<"
+    | "gt" | ">"
+    | "le" | "<="
+    | "ge" | ">="
+
+prop ::= "true"
+    | "false"
+    | "(" pprop ")"
+
+pprop ::= "old" prop
+    | "!" prop
+    | "&&" prop prop
+    | "||" prop prop
+    | "^^" prop prop
+    | "==>" prop prop
+    | "<==>" prop prop
+    | "?:" prop prop prop
+    | "let" {id}? term prop
+    | unpred prop
+    | binpred prop prop
+    | "forall" numerical_type {id}? prop
+    | "exists" numerical_type {id}? prop
+    | (* predicate application *)
+
+wasm_int_type ::= "i32"
+    | "i64"
+
+wasm_float_type ::= "f32"
+    | "f64"
+
+wasm_type ::= wasm_int_type
+    | wasm_float_type
+
+numerical_type ::= wasm_type
+    | "integer"
+    | "real"
+
+ref_type ::= "funcref"
+    | "externref"
+
+assumes_clause ::= "(" "assumes" prop ")"
+
+requires_clause ::= "(" "requires" prop ")"
+
+assigns_clause ::= "(" "assigns" {"nothing" | memarg} ")"
+
+ensures_clause ::= "(" "ensures" prop ")"
+
+simple_behavior ::= {requires_clause | assigns_clause | ensures_clause}*
+
+named_behavior ::= "(" "behavior" {assumes_clause | requires_clause | assigns_clause | ensures_clause}* ")"
+
+contract ::= "(@contract" ind simple_behavior {named_behavior}* ")"
+
+assertion ::= "(@assert" prop ")"
+
+annotation ::= contract
+    | assertion
+
+```
+
+TODO:
+- vector type
```

Comment:
I'm not sure we want to add vector type for now. I guess you'd like to have them in order to be able to have bigger numbers than `i64` and make some manipulations to check for integer overflow?

If yes, I believe it'd be better to leave vector out for now, and instead rely on an imported module providing big numbers (we can implement it separately later, either in pure Wasm, or directly in OCaml which is likely to be much easier).